S3StorageWagon/src/main/java/com/gkatzioura/maven/cloud/s3/S3StorageRepository.java

```diff
@@ -150,25 +154,39 @@ public boolean newResourceAvailable(String resourceName,long timeStamp) throws R
     }
 
 
-    public List<String> list(String path) {
+    public List<String> list(String path) throws ResourceDoesNotExistException {
 
         String key = resolveKey(path);
 
         ObjectListing objectListing = amazonS3.listObjects(new ListObjectsRequest()
                     .withBucketName(bucket)
-                    .withPrefix(key));
+                    .withPrefix(key)
+                    .withDelimiter("/"));  // makes us traverse as if using directories
+
 
         return totalObjects(objectListing);
     }
 
-    private List<String> totalObjects(ObjectListing objectListing) {
+    private List<String> totalObjects(ObjectListing objectListing) throws ResourceDoesNotExistException {
 
-        List<String> objects = new ArrayList<>();
+        int prefixLength = objectListing.getPrefix().length();
 
-        objectListing.getObjectSummaries().forEach(os->objects.add(os.getKey()));
+        if(objectListing.getObjectSummaries().isEmpty() && objectListing.getCommonPrefixes().isEmpty()) {
+            // part of the Wagon.listObjects contract
+            throw new ResourceDoesNotExistException(objectListing.getPrefix() + " is not a directory");
+        }
 
-        if(objectListing.isTruncated()) {
+        List<String> objects = Stream.concat(
+                        objectListing.getObjectSummaries()  // objects directly inside this "directory"
+                            .stream()
+                            .map(os->os.getKey().substring(prefixLength))  // ditch the prefix to return relative path
```

Comment:
This was actually the first issue I ran into when I started down this path: if I provided an S3 url with a path prefix in it, that prefix was being duplicated, because the listing operation included it as part of the file path. This strips the prefix (and the next line gets rid of the "directory objects" S3 creates if you create directories manually in the S3 interface)